### ISSUE TITLE
Teuchos::Comm: Add cheap conversion between Comm<U> and Comm<V>

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#5455]
## Bug Report
@trilinos/teuchos @cgcgcg @trilinos/thyra @trilinos/tpetra 

### Description

It should be easy and cheap to convert between `Teuchos::Comm<int>` and `Teuchos::Comm<T>` for any integer `T`.  We need this for interactions between Tpetra and Thyra, e.g., for PR https://github.com/trilinos/Trilinos/pull/5400.

**Comment by csiefer2** [comment]
We should have just deprecated that template parameter...

**Comment by kddevin** [comment]
I'm curious:  why does code exist using Comm<V> with V != int?   Teuchos::Comm with MPI works only with int, right?  
Deprecating is a good idea.  Finding someone with cycles to do it is the problem. :(

**Comment by ibaned** [comment]
I agree that we should basically `static_assert(std::is_same<T, int>::value, "anything else is crazy")` and eventually remove the template parameter. MPI uses `int`, and no alternative is anywhere on our radar. Top500 supercomputers have plateaued around supporting a maximum of one million MPI ranks and don't show many signs of approaching 2 billion MPI ranks.

**Comment by bartlettroscoe** [comment]
I `int` not also used for the size of the buffers of data being copied?  With an signed 32 bit integer, that tops off at just over 2 billion elements, no?  Is that not still an issue?

But I agree that `Teuchos::Comm<>` should not have been templated (I argued against it at the time).

**Comment by mhoemmen** [comment]
The MPI Standard is moving towards separating MPI process rank from counts and displacements.  Process ranks will stay `int`.  See https://github.com/mpi-forum/mpi-issues/issues/137 .

**Comment by ibaned** [comment]
We could just change the methods of `Teuchos::Comm` to accept a signed 64-bit integer as their count argument, and convert to whatever the underlying MPI supports as necessary internally.

**Comment by mhoemmen** [comment]
Jeff Hammond's "BigMPI" [paper](https://github.com/jeffhammond/BigMPI-paper/blob/master/exampi14_submission_3.pdf) and library shows a library-based solution.  Process ranks remain 32 bits.

**Comment by mhoemmen** [comment]
@ibaned wrote:
> We could just change the methods of `Teuchos::Comm` to accept a signed 64-bit integer as their count argument, and convert to whatever the underlying MPI supports as necessary internally.

This would be an interesting way to maintain backwards compatibility with older MPI versions, once MPI BigCount lands.

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> But I agree that `Teuchos::Comm<>` should not have been templated (I argued against it at the time).

Thank you for fighting for sanity :-) .  People forget that Comm, BLAS, and LAPACK started in the Tpetra namespace, circa 2004.

**Comment by mhoemmen** [comment]
begone autocloser
